From 1590faba19486ef897bfe60fbb5085001e2937ec Mon Sep 17 00:00:00 2001
From: Dongli Zhang <dongli.zhang@oracle.com>
Date: Tue, 14 May 2019 12:38:00 +0800
Subject: [PATCH 1/1] timer: trigger timer via pic and lapic

repo: https://github.com/finallyjustice/xv6-x86_64

This is based on commit 8700f187bac4.

The pic will trigger interrupt via lapic lint0.

Signed-off-by: Dongli Zhang <dongli.zhang@oracle.com>
---
 kernel/lapic.c | 10 ----------
 kernel/main.c  |  5 +++--
 kernel/timer.c |  4 ++--
 kernel/trap.c  |  1 +
 4 files changed, 6 insertions(+), 14 deletions(-)

diff --git a/kernel/lapic.c b/kernel/lapic.c
index ada18f4..c5b1da6 100644
--- a/kernel/lapic.c
+++ b/kernel/lapic.c
@@ -61,16 +61,6 @@ lapicinit(void)
   // Enable local APIC; set spurious interrupt vector.
   lapicw(SVR, ENABLE | (T_IRQ0 + IRQ_SPURIOUS));
 
-  // The timer repeatedly counts down at bus frequency
-  // from lapic[TICR] and then issues an interrupt.  
-  // If xv6 cared more about precise timekeeping,
-  // TICR would be calibrated using an external time source.
-  lapicw(TDCR, X1);
-  lapicw(TIMER, PERIODIC | (T_IRQ0 + IRQ_TIMER));
-  lapicw(TICR, 10000000); 
-
-  // Disable logical interrupt lines.
-  lapicw(LINT0, MASKED);
   lapicw(LINT1, MASKED);
 
   // Disable performance counter overflow interrupts
diff --git a/kernel/main.c b/kernel/main.c
index 900b632..386561f 100644
--- a/kernel/main.c
+++ b/kernel/main.c
@@ -35,8 +35,9 @@ main(void)
   fileinit();      // file table
   iinit();         // inode cache
   ideinit();       // disk
-  if(!ismp)
-    timerinit();   // uniprocessor timer
+
+  timerinit();
+
   startothers();   // start other processors
   kinit2(P2V(4*1024*1024), P2V(PHYSTOP)); // must come after startothers()
   userinit();      // first user process
diff --git a/kernel/timer.c b/kernel/timer.c
index 8df75a9..190268d 100644
--- a/kernel/timer.c
+++ b/kernel/timer.c
@@ -26,7 +26,7 @@ timerinit(void)
 {
   // Interrupt 100 times/sec.
   outb(TIMER_MODE, TIMER_SEL0 | TIMER_RATEGEN | TIMER_16BIT);
-  outb(IO_TIMER1, TIMER_DIV(100) % 256);
-  outb(IO_TIMER1, TIMER_DIV(100) / 256);
+  outb(IO_TIMER1, 255);
+  outb(IO_TIMER1, 255);
   picenable(IRQ_TIMER);
 }
diff --git a/kernel/trap.c b/kernel/trap.c
index d1b37bc..30ee173 100644
--- a/kernel/trap.c
+++ b/kernel/trap.c
@@ -50,6 +50,7 @@ trap(struct trapframe *tf)
 
   switch(tf->trapno){
   case T_IRQ0 + IRQ_TIMER:
+    cprintf("timer on cpu %d\n", cpu->id);
     if(cpu->id == 0){
       acquire(&tickslock);
       ticks++;
-- 
2.7.4

